src/rebar_dir.erl

```diff
@@ -227,7 +227,7 @@ do_make_relative_path([H|T1], [H|T2]) ->
     do_make_relative_path(T1, T2);
 do_make_relative_path(Source, Target) ->
     Base = lists:duplicate(max(length(Target) - 1, 0), ".."),
-    filename:join(Base ++ Source).
+    filename:join([Base, Source]).
```

Comment:
this just appears to be breaking every test we have.
The issue there is that `Source` and `Target` here are paths that have been already split up with `filename:split`. They are meant to be lists of directory names, which is why they are appended together with `++` and not just wrapped in a list (which would indicate that `Base/Source` is the intended result).

I think this patch does not work for what you're intending here.

Reply:
Let me check that again.

Reply:
Until I fix it, for reference, here's `Base` and `Source` when the type is unknown:
* `Base` = [],
* `Source` = [].

Reply:
Here's the output for `rebar3 dialyzer`:
```erlang
:0: Unknown type a:a/0
```

Reply:
And here's the output after the current code (which breaks stuff) is compiled and used:
```erlang
./
   0: Unknown type b:b/0
```

Reply:
@ferd, do you believe I should look "higher up" in the code to understand why we get to this point? Or having a "hack" at this place would be acceptable? [fwiw, I don't prefer the hack approach]

Reply:
(if either was `[string()]` as expected, this wouldn't break)

Reply:
if you're getting an empty path back and that we can't make an empty path relative, the fix is likely to be in the dialyzer module on our end to catch that case and display something specific about it.

The current fix just isn't valid. You're just changing the call-site where filenames are handled so that the invalid data passed in gives a reasonable-looking result, but we should just not be trying to make an empty path relative in the first place.

The distinction is just:

```erlang
1> filename:join([] ++ []).
** exception error: no function clause matching filename:join([]) (filename.erl, line 432)
2> filename:join([[], []]).
[]
```
Which shows the real problem in the full format:

```erlang
4> filename:join(["a","b"] ++ ["c","d"]).
"a/b/c/d"
5> filename:join([["a","b"], ["c","d"]]). 
"ab/cd"
```

The fix has to be at a higher level where we catch that the value we're passed won't be a valid path to normalize and pick something else. The unknown type error itself seems to just have no path, so we can pick a default (`./`) but we could very well just replace it with `?` or `<path unknown>` and be as clear.

Reply:
👍 I'll look into it, closer to the source (after `dialyzer:run/1`, in `rebar3`).

Reply:
I moved the logic closer to the source, which will produce
```erlang
/<path unknown>
   0: Unknown type b:b/0
```
(and seemingly is not affecting the tests)